GeoIP2 implementation as a plugin

sgiehl opened the pull request:
As using GeoIP2 requires a schema change on log tables to be able to track ISO region codes, GeoIP2 will be included as a new plugin. This way we can avoid an automatic schema change while updating Matomo.

This plugin will be automatically enabled for new installs. Instances updating to the new version will have the plugin included disabled. Enabling the plugin will then trigger the schema change.

As long as the GeoIP2 plugin is disabled nothing will change for Geolocation configuration, tracked data and API.

As soon as the GeoIP2 plugin is enabled some stuff changes:

**Autoupdating of GeoIP databases**
Autoupdating feature for databases will switch to configure updates for GeoIP2 databases. 
For existing installs with configured updates for GeoIP, it will still show those settings as read only. This settings will be automatically removed as soon as GeoIP2 updates are configured

**Tracking**
As soon as a GeoIP2 provider is chosen region codes will be stored as ISO codes. [GeoIP Legacy used FIPS codes]. To convert old log data from FIPS to ISO the command `usercountry:convert-region-codes` can be used. The command will convert all codes stored before the date of switching to a GeoIP2 provider. 

**Changes in API**
Until a GeoIP2 provider is selected nothing changes and still old FIPS codes and region names will be returned.
After switching to one all old codes included in archives will be converted [FIPS => ISO] if possible and region names will be used accordingly.

*Switching back from GeoIP2 to GeoIP is only possible by disabling GeoIP2 plugin. This is on purpose as it may cause problems if FIPS and ISO codes are mixed too much in the database.*

replaces #12553 
fixes #4487

sgiehl committed:
require geoip2/geoip2 composer package

sgiehl committed:
Determine region name based on Location Provider

sgiehl committed:
Adds empty GeoIp2 plugin

sgiehl committed:
move location_region column definition to GeoIp2 plugin
so it's get changed as soon as the plugin is activated

sgiehl committed:
Adds GeoIP2 location providers

sgiehl committed:
ignore GeoIP2 mmdb files

sgiehl committed:
Adds script to generate GeoIP2 test databases

sgiehl committed:
Adds Command to convert region codes from FIPS to ISO for old log table records

sgiehl committed:
Adds GeoIP2 AutoUpdater

sgiehl committed:
Use GeoIP2 in tests

sgiehl committed:
update test files

sgiehl committed:
code fixes

sgiehl committed:
adds tests

sgiehl committed:
rename old GeoIP providers to Legacy

sgiehl committed:
Let GeoIP autoupdater UI handle GeoIp2 as well

sgiehl committed:
convert region codes to ISO in API after switch to GeoIP2

sgiehl committed:
do not show GeoIP providers if GeoIP2 plugin is enabled an no GeoIP Legacy provider is still in use

sgiehl committed:
small fixes

diosmosis commented on tests/PHPUnit/Fixtures/ManyVisitsWithGeoIP.php:
```diff
@@ -58,21 +58,20 @@ class ManyVisitsWithGeoIP extends Fixture
     public function setUp()
     {
         $this->setUpWebsitesAndGoals();
-        self::downloadGeoIpDbs();
```
Are the geoip files provided in some other way during tests now? Or would I have to download them myself locally to run these tests?

diosmosis replied:
Nevermind, see the test files below. For the binary .mmdb file, would using LFS be a good idea?

sgiehl replied:
Theoretically it would be good to add them to LFS. But we currently skip downloading any LFS files on travis for all builds except the UI builds. That saves a lot time. But as the mmdb files are required for Integration and System tests, we would need to download them there as well. Also currently the mmdb files are quite small, so guess there's no need to move them to LFS.

diosmosis replied:
:+1:

diosmosis reviewed (commented):
Left some comments & suggestions, haven't tested locally yet.

sgiehl commented on plugins/GeoIp2/Commands/ConvertRegionCodesToIso.php:
```diff
@@ -0,0 +1,121 @@
+        }
+
+        // remove all entries where no change has to be done.
+        Db::query('DELETE FROM `'.Common::prefixTable(self::MAPPING_TABLE_NAME).'` WHERE `fips_code` = `iso_code`');
```
sure. will change that

Commit:
review changes

sgiehl commented on plugins/GeoIp2/LocationProvider/GeoIp2/Php.php:
```diff
@@ -0,0 +1,288 @@
+                    $result[self::CITY_NAME_KEY] = true;
+                    $result[self::POSTAL_CODE_KEY] = true;
+                    $result[self::LATITUDE_KEY] = true; // will be removed from GeoLite2-City in 2019
+                    $result[self::LONGITUDE_KEY] = true; // will be removed from GeoLite2-City in 2019
```
It will only be removed from GeoLite database published in 2019 and the paid versions will still contain such data.
Also if someone is using an old database and doesn't update after 2019 he still would have those values... Haven't spend much thought on that, but we have [an issue](https://github.com/matomo-org/matomo/issues/12411) for that. Aldo there might be some changes required for tracking lat/long at all re GDPR, so guess no need to add some checks in code for now.

diosmosis replied:
Gotcha

sgiehl commented on plugins/UserCountry/API.php:
```diff
@@ -89,33 +91,37 @@ public function getRegion($idSite, $period, $date, $segment = false)
+                        $regionCode  = getElementFromStringArray($label, $separator, 0, '');
+                        $countryCode = getElementFromStringArray($label, $separator, 1, '');
+
+                        //
```
nope

sgiehl commented on plugins/UserCountry/API.php:
```diff
@@ -155,31 +161,46 @@ public function getCity($idSite, $period, $date, $segment = false)
+                            $splitLabel[2] = strtolower($countryCode);
+                        }
+
+                        return implode($separator, $splitLabel);
```
not exactly. The values in `getCity()` also includes the city name. So it needs to be cut off and attached after converting again.

diosmosis replied:
Gotcha

sgiehl commented on plugins/UserCountry/LocationProvider.php:
```diff
@@ -117,6 +117,19 @@
+     *
+     * @return array
+     */
+    abstract public function getInfo();
```
That's on purpose. Matomo would already break if the method is not defined in any location provider, as it is be called here: https://github.com/matomo-org/matomo/blob/3.x-dev/plugins/UserCountry/LocationProvider.php#L220

diosmosis replied:
Oh I see, so it should’ve been defined before, :+1:

sgiehl commented on plugins/UserCountry/LocationProvider.php:
```diff
@@ -273,7 +304,11 @@ public static function getAllProviderInfo($newline = "\n", $includeExtra = false
+            $optionValue = Option::get(self::CURRENT_PROVIDER_OPTION_NAME);
+        } catch (\Exception $e) {
+            $optionValue = false;
+        }
```
Normally that does not occur. But some of the tests are calling that method without having the database set up.

diosmosis replied:
I see, interesting

sgiehl commented on plugins/UserCountry/API.php:
```diff
@@ -89,33 +91,37 @@ public function getRegion($idSite, $period, $date, $segment = false)
+
+                        //
+                        list($countryCode, $regionCode) = GeoIp2::convertRegionCodeToIso($countryCode,
+                            $regionCode, true);
```
I'll check

sgiehl commented on plugins/GeoIp2/LocationProvider/GeoIp2/Php.php:
```diff
@@ -0,0 +1,288 @@
+                            $result[self::REGION_CODE_KEY] = strtoupper($subdivision->isoCode);
+                            $result[self::REGION_NAME_KEY] = $subdivision->name;
+                        }
+                        break;
```
sure. why not

sgiehl commented on plugins/GeoIp2/GeoIP2AutoUpdater.php:
```diff
@@ -0,0 +1,647 @@
+        } else {
+            $ext = end(explode(basename($path), '.', 2));
+            throw new Exception(Piwik::translate('UserCountry_UnsupportedArchiveType', "'$ext'"));
+        }
```
It would be possible for tar.gz files. All other unzip classes do not have a possibility to list and extract specific content. My problem here is, that I do not know if archives of paid databases do contain the same structure and have the same archive type. The current implementation is most flexible, as it extracts everything and then checks if any extracted file matches a database name...

mattab replied:
agreed here, @sgiehl can we delete all files that don't match a white-listed extension? this would help make sure no php or other executable files could be downloaded by an attacker on the server.

diosmosis replied:
From what I can tell, the updater only supports .tar.gz & .gz files, and .gz archives can only have a single file, so filtering them out for the .tar.gz file would be enough.

sgiehl replied:
I'll check that later

sgiehl replied:
I've now changed the extraction method to only extract the first file with a valid geoip database name

Commit:
Use correct region translations

Commit:
Show correct message if no database can be found

diosmosis commented on plugins/GeoIp2/GeoIP2AutoUpdater.php:
```diff
@@ -0,0 +1,647 @@
+            if (file_exists($outputPath)) {
+                unlink($outputPath);
+            }
+            unlink($path);
```
Any thoughts @mattab? If it is useful, I think it can be added in another PR, though, so no need to worry about it now.

mattab replied:
Let's keep the existing GeoipAutoUpdater behavior for now :+1:

diosmosis commented:
Noticed there's a bug in pear/archive_tar which piwik/decompress uses which causes a fatal error when downloading databases on PHP 7: https://pear.php.net/bugs/bug.php?id=21218

I think matomo/decompress uses the fixed version.

sgiehl committed:
if log tables have been converted, use archive date to check if region codes still need to be converted to iso

diosmosis commented:
@mattab tested locally all features seem to work, except unarchiving on PHP 7. Leaving up to you whether my remaining comments should be addressed.

sgiehl committed:
fix tests

sgiehl committed:
Improves extracting GeoIP2 databases

sgiehl committed:
Adjust GeoLocation diagnostics

diosmosis commented:
@sgiehl Looked again, looks good to me except for php 7 archive tar fatal error (see my comment above for more info).

sgiehl commented:
I read it. But that seems not a "bug" related to the changes for GeoIP2, right? It should currently occur for GeoIP Legacy as well I guess.
Maybe we could create a issue to fix that later. I'm not able to handle that the next days, if you see an easy solution feel free to fix it ;)

diosmosis commented:
Easiest fix is to run `composer require pear/archive_tar` locally, I'll do this later today.

diosmosis commented:
PHP 7 bug no longer exists after 8fbe252, the offending method is no longer called. Moving back to extractInString fixed this.

fengkaijia commented on plugins/UserCountry/API.php:
```diff
@@ -89,33 +92,48 @@ public function getRegion($idSite, $period, $date, $segment = false)
                     'label',
-                    function ($label) {
-                        if ($label == '1|ti') {
-                            return '14|cn';
```
I guess this hard-coded conversion here and start from line 172 is still required for the old format.

sgiehl replied:
True. My original implementation converted all codes so it wasn't needed anymore, but with this implementation it's still needed. Will add it again

sgiehl committed:
readds old taiwan fixes

sgiehl committed:
Assume all third party location providers as 'recommended'

sgiehl commented:
I've now tested the maps with geoip2 changes_

Realtime map should work as before, as it doesn't use any region data.

The other maps will lack some information on region level. (Country and City level will still work as before) All countries where the new iso codes do not match any old fips codes the map will show those visits as "unknown region". (Note: US and Canada are not affected, as they used iso codes before, also some countries where the codes didn't change between fips and iso will still show up correctly).

To have working maps on region level with geoip2 we need to create completely new maps that also show correct regions (based in ISO) and then force all location providers to use ISO region codes in order to show region data on the map

diosmosis commented on plugins/GeoIp2/Commands/ConvertRegionCodesToIso.php:
```diff
@@ -0,0 +1,122 @@
+                    continue; // nothing needs to be changed, so ignore the mapping
+                }
+
+                Db::query('INSERT INTO `'.Common::prefixTable(self::MAPPING_TABLE_NAME).'` VALUES (?, ?, ?)', [$country, $fips, $iso]);
```
Small comment, but think this should be `INSERT IGNORE` in case the command fails for some unknown reason and the user tries it again. w/o it, the command will fail here in such a situation.

Alternatively, could drop the table above before trying to create it.

sgiehl replied:
I'll add a drop

diosmosis commented on plugins/GeoIp2/Commands/ConvertRegionCodesToIso.php:
```diff
@@ -0,0 +1,121 @@
+    }
+
+
+}
```
@mattab suggested printing out the SQL as it's running

mattab committed:
Download database over HTTPS

mattab committed:
remove outdated comment (see https://github.com/matomo-org/matomo/issues/12411)

mattab committed:
Remove indication that Geoip2 may be slow, since we found it should be quite fast and this should not be an issue in theory

mattab commented:
Feedback @sg

* Wizard: maybe we should simply remove the downloaded file if anything fails to download or extract
* Once the wizard has been used, and the settings saved, we should reload the page, so that it says "Geoip2: installed". otherwise use has to manually refresh the page to actually enable geoip2 which is not obvious.
* Region may be incorrectly detected. For example for IP address 194.57.91.215 we get the following "Location" result in the [geoip2 demo](https://www.maxmind.com/en/geoip-demo): `Besançon, Doubs, Bourgogne-Franche-Comte,  France, Europe`. The region name is  `Bourgogne-Franche-Comte`. However in our test files we can see that this IP was sset to the region of `Doubs` which is not the region but the `département` (there are several departements in each region). -> this seems a big issue/regression that (can be seen in the test output)
* Let's update the  `Region.php` to define the segment & set the accepted values as ISO code? Currently they read `protected $acceptValues = '01 02, OR, P8, etc.<br/>eg. region=A1;country=fr';`
* Getting the following error after enabling Geoip2 `The value "" is not a valid IP address.` -> as discussed on slack, maybe we should overwrite that method for geoip2 provider, and always use ip from header, without trying to convert it to ipv4 if possible?

sgiehl committed:
skip detection if IP is empty & do not try convert IP to IPv4

sgiehl committed:
remove downloaded file if an error occurs while extracting

sgiehl committed:
command should be runnable multiple times

sgiehl committed:
use ISO codes for suggested region codes

sgiehl committed:
reload after wizard success

sgiehl commented:
> Wizard: maybe we should simply remove the downloaded file if anything fails to download or extract

it's not automatically removed when extracting fails

> Once the wizard has been used, and the settings saved, we should reload the page, so that it says "Geoip2: installed". otherwise use has to manually refresh the page to actually enable geoip2 which is not obvious.

changed.

> Region may be incorrectly detected. For example for IP address 194.57.91.215 we get the following "Location" result in the geoip2 demo: Besançon, Doubs, Bourgogne-Franche-Comte, France, Europe. The region name is Bourgogne-Franche-Comte. However in our test files we can see that this IP was sset to the region of Doubs which is not the region but the département (there are several departements in each region). -> this seems a big issue/regression that (can be seen in the test output)

GeoIP2 returns multiple regions. Currently I'm using the smallest known region that is detected. We could use the largest region detected, which would "fix" your regression for France. BUT it would cause changes for regions in UK. The first regions reported would be England, Ireland, Schottland. 

For example the result for `212.44.43.3` for GeoIP Legacy is:
```
country_code,country_name,city_name,region_code,region_name
GB,United Kingdom,Linton,C3,Cambridgeshire
```
For GeoIP 2 using the first region it would be 
```
country_code,country_name,city_name,region_code,region_name
GB,United Kingdom,Linton,ENG,England
```
So not sure how to handle that best... @mattab  any thoughts? shall we use the smallest or the biggest detected region?

> Let's update the Region.php to define the segment & set the accepted values as ISO code? Currently they read protected $acceptValues = '01 02, OR, P8, etc.<br/>eg. region=A1;country=fr';

changed

> Getting the following error after enabling Geoip2 The value "" is not a valid IP address. -> as discussed on slack, maybe we should overwrite that method for geoip2 provider, and always use ip from header, without trying to convert it to ipv4 if possible?

That should be fixed now.

diosmosis committed:
Drop table if exists.

diosmosis committed:
Fix two translation keys.

diosmosis commented:
Fixed two minor issues I found, other than that looks good. Couple things I couldn't test:

* test w/ nginx geoip2 module (didn't have time to configure the module)
* test w/ Org/ISP databases

diosmosis commented:
Looks like there are a couple system test failures. Once those are fixed I'd say it's ready to merge (after updating screenshots).

mattab commented:
@sgiehl I'm not sure about this, but it seems like it could break a lot... it's quite important to get the region right.. 

-> Could you maybe compare what we used to have, to the new format, for the top 50 countries or so? Ideally we have a mapping where we know which region level we need to use for a given country, so it's correct for FR, UK, USA, and other 50+ countries... 

@diosmosis just sent you the commercial ISP DB for testing

@sgiehl could you please apply pending feedback (test, UI files) and I'll proceed to merge the PR, ideally in 12 hours or so :+1:

diosmosis commented:
@mattab tested w/ isp db, no issues found

sgiehl committed:
add special region handling for UK

sgiehl committed:
update system test

sgiehl commented:
> -> Could you maybe compare what we used to have, to the new format, for the top 50 countries or so? Ideally we have a mapping where we know which region level we need to use for a given country, so it's correct for FR, UK, USA, and other 50+ countries...

There is no good way to check if it's the same. For almost all countries the region codes and names changed anyway by using ISO instead of FIPS.
For now I've added an exception for UK, so it uses the smallest region in this case. For all other countries it will now use the biggest region.

sgiehl committed:
update UI files

sgiehl committed:
submodule update

sgiehl committed:
update test files
